coderd/workspaceapps/auth.go

```diff
@@ -0,0 +1,501 @@
+	"github.com/coder/coder/coderd/rbac"
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/site"
+)
+
+const (
+	// TODO: configurable expiry
+	TicketExpiry = time.Minute
+
+	// RedirectURIQueryParam is the query param for the app URL to be passed
+	// back to the API auth endpoint on the main access URL.
+	RedirectURIQueryParam = "redirect_uri"
+)
+
+// ResolveRequest takes an app request, checks if it's valid and authenticated,
+// and returns a ticket with details about the app.
+//
+// The ticket is written as a signed JWT into a cookie and will be automatically
+// used in the next request to the same app to avoid database calls.
+//
+// Upstream code should avoid any database calls ever.
+func (p *Provider) ResolveRequest(rw http.ResponseWriter, r *http.Request, appReq Request) (*Ticket, bool) {
+	err := appReq.Validate()
+	if err != nil {
+		p.writeWorkspaceApp500(rw, r, &appReq, err, "invalid app request")
+		return nil, false
+	}
+
+	if appReq.WorkspaceAndAgent != "" {
+		// workspace.agent
+		workspaceAndAgent := strings.SplitN(appReq.WorkspaceAndAgent, ".", 2)
+		appReq.WorkspaceAndAgent = ""
+		appReq.WorkspaceNameOrID = workspaceAndAgent[0]
+		if len(workspaceAndAgent) > 1 {
+			appReq.AgentNameOrID = workspaceAndAgent[1]
+		}
+
+		// Sanity check.
+		err := appReq.Validate()
+		if err != nil {
+			p.writeWorkspaceApp500(rw, r, &appReq, err, "invalid app request")
+			return nil, false
+		}
+	}
+
+	// Get the existing ticket from the request.
+	ticketCookie, err := r.Cookie(codersdk.DevURLSessionTicketCookie)
+	if err == nil {
+		ticket, err := p.ParseTicket(ticketCookie.Value)
+		if err == nil {
+			if ticket.MatchesRequest(appReq) {
+				// The request has a ticket, which is a valid ticket signed by
+				// us, and matches the app that the user was trying to access.
+				return &ticket, true
+			}
+		}
+	}
+
+	// There's no ticket or it's invalid, so we need to check auth using the
+	// session token, validate auth and access to the app, then generate a new
+	// ticket.
+	//
+	// We use the regular API key extraction middleware here to avoid any
+	// differences in behavior between the two.
+	var (
+		ticket = Ticket{
+			AccessMethod:      appReq.AccessMethod,
+			UsernameOrID:      appReq.UsernameOrID,
+			WorkspaceNameOrID: appReq.WorkspaceNameOrID,
+			AgentNameOrID:     appReq.AgentNameOrID,
+			AppSlugOrPort:     appReq.AppSlugOrPort,
+		}
+		ticketOK = false
+	)
+	httpmw.ExtractAPIKey(httpmw.ExtractAPIKeyConfig{
+		DB:            p.Database,
+		OAuth2Configs: p.OAuth2Configs,
+		// Optional is true to allow for public apps. If an authorization check
+		// fails and the user is not authenticated, they will be redirected to
+		// the login page below.
+		RedirectToLogin:             false,
+		DisableSessionExpiryRefresh: p.DeploymentConfig.DisableSessionExpiryRefresh.Value,
+		Optional:                    true,
+	})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I can't say I'm a huge fan of seeing this pattern cropping up. Can't we refactor it so that we aren't having to call middlewares outside of a router?

Reply:
The middleware makes database calls, which we're trying to avoid so I have to call it after checking the ticket. I agree it's not very good, but I didn't want to have to refactor the middleware as it'd make the PR another 1k lines...

The middleware is pretty complicated too so we can't reimplement the parts we need here either without making the handler much larger